theme/gallery-album.tsx

```diff
@@ -0,0 +1,29 @@
+import type { Culture, NewsroomGallery } from '@prezly/sdk';
+
+import { api } from '@/theme-kit';
+
+interface Match {
+    locale: Culture['code'];
+    uuid: string;
+}
+
+export async function match({ uuid }: Match) {
+    const { contentDelivery } = api();
+
+    const gallery = await contentDelivery.gallery(uuid);
+
+    return gallery && { gallery };
+}
```

Comment:
Theme page modules can export a `match` function to additionally process the router match params to:

- validate it against the real API data
- enrich the page props with real data objects (to not request them again)

This is necessary to make multiple same-level dynamic routes work without collisions: primarily for `/[locale]` and `/[slug]` routes.

Reply:
This particular case doesn't really utilize the server components paradigm, where the data is fetched at the component level. In this current form it looks more like the old `getServerSideProps` 🤔

Reply:
> In this current form it looks more like the old getServerSideProps 🤔

You're right that it's similar. But it's also a way to reduce duplication.

An advantage is that this only fetches the route root object (the `gallery` in this case), allowing the underlying components to fetch more data as they need. For example, the `newsroom` object will never be passed down the tree and can be requested via `await contentDelivery.newsroom()` anywhere. Same for `categories`, language boilerplate settings, and so on.

Reply:
Still, apart from the `/[locale]` - `/[slug]` conflict, I don't see any benefit of this approach over simply fetching the data in the component and calling `notFound` if the gallery is not found.

Also, with this approach, for every request to any route, this data-fetching code will run for every defined route, which is totally unnecessary and introduces extra load on the API without real need for it.

I get that it makes sense for routes that can conflict and should rely on API to resolve, but for any other routes that have a clear static part (e.g. `/s/[uuid]`, `/media/album/[uuid]`, `/search`), there is no point in calling the data-fetching logic until the route is actually matched and rendered..

Reply:
> but for any other routes that have a clear static part (e.g. /s/[uuid], /media/album/[uuid], /search), there is no point in calling the data-fetching logic until the route is actually matched and rendered.

But it will not. The module-level `match()` function is only called if the static route pattern has already matched, and skipped if it does not. So, in fact, this will only make a different for the conflicting routes.

Reply:
Okay, that it clears it up a bit. 
But still, it's a bit confusing to me, that there's a function called `match` whose primary function is to help the router match the routes that could conflict, but then it somehow also provides data for the page component.

If you're worried about the same request being called twice, Next.js caching should take care of that (since the same function would be called with the same arguments). I agree that `match` is useful, but I'd leave it to only serve as the helper for the router, not a data-fetching method.

Reply:
> If you're worried about the same request being called twice, Next.js caching should take care of that (since the same function would be called with the same arguments)

Yeah, I know. That's why I said that [newsroom object will never be passed down](https://github.com/prezly/theme-nextjs-bea/pull/912#discussion_r1358233949) and can be fetched wherever it is needed.

> but I'd leave it to only serve as the helper for the router, not a data-fetching method.

If we don't allow reusing the data-fetching result of the `match()` function, then you'll need to fetch it again and check if it's not null again:

```
export function match({ slug }) {
  const story = api.story({ slug });
  return Boolean(story);
}

export default function Page({ slug }) {
  const story = api.story({ slug });

  if (!story) {
     notFound();
  }

  return <Markup story={story} />
}
```

And you're right that moving it to the render function _in other modules_ is possible (and may be preferrable). But that's unimportant. I see it as a nice side-effect of route conflict resolution logic that allows me do skip throwing `notFound()` manually in the render component. But this is just an implementation detail.